File: Python/34.find-first-and-last-position-of-element-in-sorted-array.py
#
# @lc app=leetcode id=34 lang=python
#
# [34] Find First and Last Position of Element in Sorted Array
#
# https://leetcode.com/problems/find-first-and-last-position-of-element-in-sorted-array/description/
#
# algorithms
# Medium (34.48%)
# Likes:    2177
# Dislikes: 100
# Total Accepted:    372.8K
# Total Submissions: 1.1M
# Testcase Example:  '[5,7,7,8,8,10]\n8'
#
# Given an array of integers nums sorted in ascending order, find the starting
# and ending position of a given target value.
# 
# Your algorithm's runtime complexity must be in the order of O(log n).
# 
# If the target is not found in the array, return [-1, -1].
# 
# Example 1:
# 
# 
# Input: nums = [5,7,7,8,8,10], target = 8
# Output: [3,4]
# 
# Example 2:
# 
# 
# Input: nums = [5,7,7,8,8,10], target = 6
# Output: [-1,-1]
# 
#

# @lc code=start
class Solution(object):
    def searchRange(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """

        if not nums:
            return [-1, -1]
            
        left, right = 0, len(nums) - 1
        while left < right:
            mid = left + (right - left) // 2
            if nums[mid] >= target:
                right = mid
            else:
                left = mid + 1
        if nums[left] != target:
            return [-1, -1]
        
        ans = [left]
        left, right = 0, len(nums) - 1
        while left < right:
            mid = right - (right - left) // 2
            if nums[mid] <= target:
                left = mid
            else:
                right = mid - 1
        ans.append(right)
        return ans
# @lc code=end

sol = Solution()
nums = [5,7,7,8,8,10]
target = 6
print(sol.searchRange(nums, target))

File: C++/34.find-first-and-last-position-of-element-in-sorted-array.cpp
/*
 * @lc app=leetcode.cn id=34 lang=cpp
 *
 * [34] Find First and Last Position of Element in Sorted Array
 */
#include<vector>
using namespace std;

// @lc code=start
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        // int left=0, right=nums.size()-1;
        // while (left<=right) {
        //     int mid = left + (right-left)/2;
        //     if (nums[mid]<target)
        //         left = mid+1;
        //     else if (nums[mid]>target)
        //         right = mid-1;
        //     else {
        //         left=mid, right=mid;
        //         while (left>=0 && nums[left]==target)
        //             left--;
        //         while (right<nums.size() && nums[right]==target)
        //             right++;
        //         return vector<int>{left+1, right-1};
        //     }
        // }
        // return vector<int>{-1, -1};

        vector<int> res{-1, -1};
        if (nums.empty()) return res;
        
        int l=0, r=nums.size()-1;
        while (l<r) {
            int mid = l + (r-l)/2;
            if (nums[mid]<target)
                l = mid+1;
            else
                r = mid;
        }
        if (nums[l]!=target) return res;
        
        res[0] = l;
        r=nums.size()-1;
        while (l<r) {
            int mid = l + (r-l)/2 + 1;
            if (nums[mid]>target)
                r = mid-1;
            else
                l = mid;
        }
        res[1] = r;
        return res;
    }
};
// @lc code=end

